src/webnn_native/dml/GraphDML.cpp

```diff
@@ -888,7 +947,79 @@ namespace webnn_native::dml {
+        DAWN_ASSERT(mExpression.find(inputsOperand[0].Get()) != mExpression.end());
+        ::dml::Expression input = mExpression.at(inputsOperand[0].Get());
+        DAWN_ASSERT(mExpression.find(inputsOperand[1].Get()) != mExpression.end());
+        ::dml::Expression filter = mExpression.at(inputsOperand[1].Get());
+
+        const ConvTranspose2dOptions* options = convTranspose2d->GetOptions();
+        if (options->inputLayout == wnn::InputOperandLayout::Nhwc) {
+            input = ReinterpretInputLayout(NhwcToNchw, input);
+        }
+        if (options->filterLayout != wnn::ConvTranspose2dFilterOperandLayout::Iohw) {
+            filter = ReinterpretFilterLayoutAsIohw(options->filterLayout, filter);
+        }
+        ::dml::Span<const uint32_t> strides(reinterpret_cast<const uint32_t*>(options->strides),
+                                            options->stridesCount);
+        ::dml::Span<const uint32_t> dilations(reinterpret_cast<const uint32_t*>(options->dilations),
+                                              options->dilationsCount);
+        ::dml::Span<const uint32_t> outputPadding(
+            reinterpret_cast<const uint32_t*>(options->outputPadding), options->outputPaddingCount);
+
+        ::dml::TensorDimensions inputDims = input.GetOutputDesc().sizes;
+        ::dml::TensorDimensions filterDims = filter.GetOutputDesc().sizes;
+        auto padding = options->autoPad == wnn::AutoPad::Explicit
+                           ? ExplicitPadding<ConvTranspose2dOptions>(options)
+                           : ImplicitPaddingforConvTranpose2d(options, inputDims, filterDims);
+        // dml::Span just holds the refernces, need a variable to hold the memory.
+        std::vector<const uint32_t> startPaddingVector = {padding[0], padding[2]};
+        ::dml::Span<const uint32_t> startPadding(startPaddingVector);
+        std::vector<const uint32_t> endPaddingVector = {padding[1], padding[3]};
+        ::dml::Span<const uint32_t> endPadding(endPaddingVector);
+
+        ::dml::TensorDimensions outputShape = {};
+        outputShape.push_back(inputDims[0]);   // N
+        outputShape.push_back(filterDims[1]);  // C
+        uint32_t outputHeight, outputWidth;
+        if (options->outputSizes != nullptr) {
+            outputHeight = options->outputSizes[0];
+            outputWidth = options->outputSizes[1];
+        } else {
+            outputHeight = strides[0] * (inputDims[2] - 1) + outputPadding[0] +
+                           ((filterDims[2] - 1) * dilations[0] + 1) - startPaddingVector[0] -
+                           endPaddingVector[0];
+            outputWidth = strides[1] * (inputDims[3] - 1) + outputPadding[1] +
+                          ((filterDims[3] - 1) * dilations[1] + 1) - startPaddingVector[1] -
+                          endPaddingVector[1];
+        }
+        outputShape.push_back(outputHeight);
+        outputShape.push_back(outputWidth);
```

Comment:
Could we use the calculated output shape from ConvTranspose2d.cpp?

Reply:
Yes, I have optimized the code. PTAL, thanks!